gadgets/deadlock/go/program.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"strconv"
+	"strings"
+
+	"deadlock/digraph"
+
+	api "github.com/inspektor-gadget/inspektor-gadget/wasmapi/go"
+)
+
+var g *digraph.DiGraph
+
+//export gadgetInit
+func gadgetInit() int {
+	g = digraph.NewDiGraph()
+	ds, err := api.GetDataSource("deadlock")
+	if err != nil {
+		api.Warnf("failed to get datasource: %s", err)
+		return 1
+	}
+	mutex1F, err := ds.GetField("mutex1")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mutex2F, err := ds.GetField("mutex2")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	tidF, err := ds.GetField("tid")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	pidF, err := ds.GetField("pid")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mutex1StackF, err := ds.GetField("mutex1_stack_id")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mutex2StackF, err := ds.GetField("mutex2_stack_id")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	commF, err := ds.GetField("comm")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	ds.Subscribe(func(source api.DataSource, data api.Data) {
+		mutex1, err := mutex1F.Uint64(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		mutex2, err := mutex2F.Uint64(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		tid, err := tidF.Uint32(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		pid, err := pidF.Uint32(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		mutex1Stack, err := mutex1StackF.Uint64(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		mutex2Stack, err := mutex2StackF.Uint64(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		comm, err := commF.String(data)
+		if err != nil {
+			api.Warnf("failed to get field: %s", err)
+			return
+		}
+		g.AddEdge(mutex1, mutex2, map[string]interface{}{
+			"tid":             tid,
+			"pid":             pid,
+			"mutex1_stack_id": mutex1Stack,
+			"mutex2_stack_id": mutex2Stack,
+			"comm":            comm,
+		})
+	}, 0)
+	return 0
+}
+
+//export gadgetStop
+func gadgetStop() int {
+	cycles := digraph.FindCycles(g)
```

Comment:
Presumably yes. As @burak-ok mentioned, with many edges, we get `ERRO[0023] stopping operator "wasm": calling gadgetStop: module closed with context deadline exceeded` with the current setup.

Reply:
Maybe for gadgets with `GADGET_MAPITER`, we need a way to register a callback to be called one time for every batch of data (e.g. 1 time per second).

Also, I would be curious to see profiling to understand why it is expensive and if it can be optimized.